[git-webkit] Add screen-reader friendly review wizard (Part 2)

JonWBedard opened the pull request:
#### 74313d0521aae7fbcd8c8a287b92f924f34ce34c
<pre>
[git-webkit] Add screen-reader friendly review wizard (Part 2)
<a href="https://bugs.webkit.org/show_bug.cgi?id=261242">https://bugs.webkit.org/show_bug.cgi?id=261242</a>
<a href="https://rdar.apple.com/115083100">rdar://115083100</a>

Reviewed by Andres Gonzalez and Elliott Williams.

Allow user to review and comment on a pull request by editing a local file.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py:
(Review.truncate_strs): Turn a nested dictionary of strings into a nested dictionary
of lists of string based on where newline are in strings.
(Review.user_delta): Detect which users have been added or removed from a list of users.
(Review.invoke_wizard): Diff the original file with a user&apos;s local edits and return those
local edits to be converted into actions on a pull request.
(Review.main): Given the difference returned by invoke_wizard, programatically add comments,
change pull request metadata and approve (or reject) the pull request.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/review_unittest.py:
(TestReview.editor_callback):
(TestReview.editor_callback.callback):
(TestReview.test_truncate_strs):
(TestReview.test_user_delta_github):
(TestReview.test_user_delta_bitbucket):
(TestReview.test_edit_metadata):
(TestReview.test_comment_commit_message):
(TestReview.test_comment):
(TestReview.test_comment_reply):
(TestReview.test_comment_reply_only_one):
(TestReview.test_comment_reply_both):
(TestReview.test_diff_file_comment):
(TestReview.test_diff_inline_comment):
(TestReview.test_github_read_comments):
(TestReview.test_bitbucket_no_edit):
(TestReview.test_github_no_edit):
(TestReview.test_bitbucket_approve):
(TestReview.test_github_deny):
(TestReview.test_bitbucket_edit_deny):
(TestReview.test_github_edit_approve):
(TestReview.test_bitbucket_edit_refresh):
(TestReview.test_github_edit_refresh):
(TestReview.test_bitbucket_edit_comment):
(TestReview.test_github_edit_comment):
(TestReview.test_bitbucket_edit_comment_commit_message):
(TestReview.test_github_edit_comment_commit_message):
(TestReview.test_bitbucket_edit_comment_inline_diff):
(TestReview.test_github_edit_comment_inline_diff):
(TestReview.test_bitbucket_edit_title):
(TestReview.test_github_edit_title):
(TestReview.test_bitbucket_edit_open):
(TestReview.test_github_edit_close):
(TestReview.test_bitbucket_edit_merged_close):
(TestReview.test_github_edit_merged_open):
(TestReview.test_bitbucket_add_labels):
(TestReview.test_github_edit_labels):
(TestReview.test_github_edit_invalid_labels):

Canonical link: <a href="https://commits.webkit.org/279255@main">https://commits.webkit.org/279255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d8839c1a3c3b76e148d5d561889f2f149939a7a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52919 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5406 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3642 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55224 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3367 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/2344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55017 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29962 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45678 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24049 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/52763 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2998 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1801 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3154 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57791 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28060 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3120 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/52926 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45895 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7775 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29034 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py:
```diff
@@ -232,4 +395,108 @@ def main(cls, args, repository, **kwargs):
+
+        if pull_request.opened and did_approve is None and user and pull_request.author != user:
+            did_approve = dict(Yes=True, Block=False).get(Terminal.choose(
+                'Would you like approve this change?',
```
grammar: "Would you like *to* approve this change?"

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py:
```diff
@@ -232,4 +395,108 @@ def main(cls, args, repository, **kwargs):
+        if pull_request.opened and did_approve is None and user and pull_request.author != user:
+            did_approve = dict(Yes=True, Block=False).get(Terminal.choose(
+                'Would you like approve this change?',
+                default='No', options=('Yes', 'No', 'Block'),
```
Yes / No / Block does not feel very intuitive to me. In response to the question "Would you like to approve this change?", "No" sounds like r-, but it's actually what GitHub calls "Comment".

Could we use the same idioms as GitHub? Like: "How would you like to review this change?" with choices "Comment only", "Approve", "Request changes"

emw-apple reviewed (approved):
Nothing too surprising in the code. I did notice that we're inconsistent about whether PRs are called "pull requests" or "pull-requests". GitHub uses the former, and I think that's the generally accepted spelling. Would be nice to clean that up before landing.

JonWBedard committed:
[git-webkit] Add screen-reader friendly review wizard (Part 2)
https://bugs.webkit.org/show_bug.cgi?id=261242
rdar://115083100

Reviewed by Andres Gonzalez and Elliott Williams.

Allow user to review and comment on a pull request by editing a local file.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py:
(Review.truncate_strs): Turn a nested dictionary of strings into a nested dictionary
of lists of string based on where newline are in strings.
(Review.user_delta): Detect which users have been added or removed from a list of users.
(Review.invoke_wizard): Diff the original file with a user's local edits and return those
local edits to be converted into actions on a pull request.
(Review.main): Given the difference returned by invoke_wizard, programatically add comments,
change pull request metadata and approve (or reject) the pull request.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/review_unittest.py:
(TestReview.editor_callback):
(TestReview.editor_callback.callback):
(TestReview.test_truncate_strs):
(TestReview.test_user_delta_github):
(TestReview.test_user_delta_bitbucket):
(TestReview.test_edit_metadata):
(TestReview.test_comment_commit_message):
(TestReview.test_comment):
(TestReview.test_comment_reply):
(TestReview.test_comment_reply_only_one):
(TestReview.test_comment_reply_both):
(TestReview.test_diff_file_comment):
(TestReview.test_diff_inline_comment):
(TestReview.test_github_read_comments):
(TestReview.test_bitbucket_no_edit):
(TestReview.test_github_no_edit):
(TestReview.test_bitbucket_approve):
(TestReview.test_github_deny):
(TestReview.test_bitbucket_edit_deny):
(TestReview.test_github_edit_approve):
(TestReview.test_bitbucket_edit_refresh):
(TestReview.test_github_edit_refresh):
(TestReview.test_bitbucket_edit_comment):
(TestReview.test_github_edit_comment):
(TestReview.test_bitbucket_edit_comment_commit_message):
(TestReview.test_github_edit_comment_commit_message):
(TestReview.test_bitbucket_edit_comment_inline_diff):
(TestReview.test_github_edit_comment_inline_diff):
(TestReview.test_bitbucket_edit_title):
(TestReview.test_github_edit_title):
(TestReview.test_bitbucket_edit_open):
(TestReview.test_github_edit_close):
(TestReview.test_bitbucket_edit_merged_close):
(TestReview.test_github_edit_merged_open):
(TestReview.test_bitbucket_add_labels):
(TestReview.test_github_edit_labels):
(TestReview.test_github_edit_invalid_labels):

Canonical link: https://commits.webkit.org/279255@main
